### DISCUSSION TITLE
Mastermind Game on Agoric

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#10013] [Show and tell]
## Context

The purpose of this discussion is to showcase a simple dApp I have built to explore the gaming use case on the Agoric chain and improve my frontend skills.
Note: the release of the [UI Tutorial](https://docs.agoric.com/guides/getting-started/ui-tutorial/) was the trigger that motivated me to build the UI for this dApp.

**GitHub Repository**: https://github.com/Jorge-Lopes/mastermind

## Project Overview

This dApp implements a version of the Mastermind game, where the contract will act as the code maker, which will generate a sequence of four colored pegs from a set of 6 available colors. The sequence (secret code) is kept hidden from the user until it wins or loses the game. 
The user, code breaker, is allowed to make a guess of four colored pegs in a specific order and will receive a feedback in return.
The game is finished when the codebreaker guesses the exact sequence, or if it cannot guess the code within the given 10 attempts.

**Key Features:**

- Create New Game: Users can create new games by pressing the respective button.
- Guess The Code: Users can submit guesses, where each guess is checked against the secret code.
- Feedback Mechanism: For each guess, feedback is provided:
    - Black Color: Represents correct values in the correct position.
    - Red Color: Represents correct values in the wrong position.
- Game Phases: The game phase can be either:
    - Active: The game is in progress, and users can continue making guesses.
    - Completed: The user has won the game by guessing the correct code.
    - Failed: The user has lost the game by exhausting all attempts without guessing the correct code.

![screenshot](https://github.com/Jorge-Lopes/mastermind/blob/main/Mastermind_Screenshot.png)

## Challenges

**Could a game like this be monetised?**

The secret code of each new game is generated via the [Mersenne-Twister](https://www.npmjs.com/package/mersenne-twister) package, which is a pseudorandom number generator.

At the [generateSecretCode](https://github.com/Jorge-Lopes/mastermind/blob/main/contract/src/utils.js#L4-L13) function, every time a new game is initiated, a new instance of the Mersenne Twister generator is created and used to generate a 4-digit array, using the random method to produce a number between 0 and 5. 

To create the Mersenne Twister generator, we need to pass an integer (seed) into the constructor, which will always produce the same random sequence.
In this implementation, the seed being used is the absValue returned by the `getCurrentTimestamp` method of the `chainTimerService`, at the moment the `makeGameHandle` is executed.

I am questioning if a player with access to the chain logs, (e.g. by running a node) would be able to deduce the timestamp used as seed and this way predict its secretCode?

Chain logs example:
```bash
agd-1  | 2024-09-03T11:16:53.614Z SwingSet: vat: v43: walletFactory.fromBridge: { blockHeight: 1223, blockTime: 1725362212, owner: 'agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk', spendAction: '{"body":"#{\\"method\\":\\"executeOffer\\",\\"offer\\":{\\"id\\":\\"start_game_1725362197578\\",\\"invitationSpec\\":{\\"instance\\":\\"$0.Alleged: InstanceHandle#board01998\\",\\"publicInvitationMaker\\":\\"makeGameInvitation\\",\\"source\\":\\"contract\\"},\\"offerArgs\\":\\"#undefined\\",\\"proposal\\":{}}}","slots":["board01998"]}', type: 'WALLET_SPEND_ACTION' }
agd-1  | 2024-09-03T11:16:53.617Z SwingSet: vat: v43: walletFactory: { wallet: Object [Alleged: SmartWallet self] {}, actionCapData: { body: '#{"method":"executeOffer","offer":{"id":"start_game_1725362197578","invitationSpec":{"instance":"$0.Alleged: InstanceHandle#board01998","publicInvitationMaker":"makeGameInvitation","source":"contract"},"offerArgs":"#undefined","proposal":{}}}', slots: [ 'board01998' ] } }
agd-1  | 2024-09-03T11:16:53.643Z SwingSet: vat: v43: ----- SmrtWlt.4  2 getLiveOfferPayments adding store for agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk
agd-1  | 2024-09-03T11:16:53.652Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk starting executeOffer start_game_1725362197578
agd-1  | 2024-09-03T11:16:53.920Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk start_game_1725362197578 seated
agd-1  | 2024-09-03T11:16:54.123Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk offerStatus { id: 'start_game_1725362197578', invitationSpec: { instance: Object [Alleged: InstanceHandle] {}, publicInvitationMaker: 'makeGameInvitation', source: 'contract' }, offerArgs: undefined, proposal: {}, result: 'UNPUBLISHED' }
agd-1  | 2024-09-03T11:16:54.128Z SwingSet: vat: v43: wallet agoric1rwwley550k9mmk6uq6mm6z4udrg8kyuyvfszjk recording pathMap { game: 'published.mastermind.game0' }
```

### Proposed solution

While discussing this subject during office hours with @dckc , one possible solution that was raised is to take advantage of the Agoric [Orchestration](https://docs.agoric.com/guides/orchestration/) pkg to interact with some Cosmos chains that provide reliable randomness, for example [Nois](https://nois.network/).

To be detailed … 

--- 

Hope you have found this project interesting, I'm open to any feedback or suggestions you might have on this topic.
Thank you all for your time.

**Comment by dckc** [comment]
@Jovonni suggests maybe Pedersen Commitments...

or zk circuits using something like [01js](https://github.com/o1-labs/o1js). It would likely add an off-chain component for the prover.
